test/fake_http_server.cc

```diff
@@ -87,38 +98,73 @@ void FakeServer::TerminateAllStreams() {
                                      Server::connection_ptr connection) {
-  auto stream_it = path_streams.find(request.destination);
-  if (stream_it != path_streams.end()) {
-    auto& stream = stream_it->second;
-    connection->set_status(Server::connection::ok);
-    connection->set_headers(std::map<std::string, std::string>({
-        {"Content-Type", "text/plain"},
-    }));
-
-    // Create a queue for this watcher and add to the stream.
-    std::queue<std::string> my_queue;
-    stream.AddQueue(&my_queue);
-
-    // For every new string on my queue, send to the client.  The
-    // empty string indicates that we should terminate the stream.
-    while (true) {
-      std::string s = stream.GetNextResponse(&my_queue);
-      if (s.empty()) {
-        break;
+  if (request.method == "GET") {
+    auto stream_it = path_streams.find(request.destination);
+    if (stream_it != path_streams.end()) {
+      auto& stream = stream_it->second;
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
+          {"Content-Type", "text/plain"},
+      }));
+
+      // Create a queue for this watcher and add to the stream.
+      std::queue<std::string> my_queue;
+      stream.AddQueue(&my_queue);
+
+      // For every new string on my queue, send to the client.  The
+      // empty string indicates that we should terminate the stream.
+      while (true) {
+        std::string s = stream.GetNextResponse(&my_queue);
+        if (s.empty()) {
+          break;
+        }
+        connection->write(s);
       }
-      connection->write(s);
+      return;
+    }
+
+    auto it = get_handlers.find(request.destination);
+    if (it != get_handlers.end()) {
+      const auto& handler = it->second;
+      std::map<std::string, std::string> headers;
+      for (const auto& h : request.headers) {
+        headers[h.name] = h.value;
+      }
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
+          {"Content-Type", "text/plain"},
+      }));
+      connection->write(handler(request.destination, headers));
+      return;
     }
-    return;
   }
 
-  auto it = path_responses.find(request.destination);
-  if (it != path_responses.end()) {
-    connection->set_status(Server::connection::ok);
-    connection->set_headers(std::map<std::string, std::string>({
-        {"Content-Type", "text/plain"},
-    }));
-    connection->write(it->second);
-    return;
+  if (request.method == "POST") {
+    auto it = post_handlers.find(request.destination);
+    if (it != post_handlers.end()) {
+      const auto& handler = it->second;
+      std::map<std::string, std::string> headers;
+      for (const auto& h : request.headers) {
+        headers[h.name] = h.value;
+      }
+      std::string body;
+      std::mutex mutex;
+      mutex.lock();
+      connection->read([&](Server::connection::input_range range,
```

Comment:
[Optional] Fun fact: at least in GCC 5.4 (Xenial) and GCC 7.3 (Stretch), the following works as well:
```c++
      connection->read([&](Server::connection::input_range range, ...) {
        std::unique_lock<std::mutex> done(std::move(body_retrieved));
        body.append(range.begin(), range.end());
      });
```
Sadly, it doesn't work with GCC 4.8 (Trusty, CentOS 7). :-(
```
/usr/include/c++/4.8/functional:2071:3: error: cannot pass objects of non-trivially-copyable type 'class std::shared_ptr<boost::network::http::async_connection<boost::network::http::tags::http_server, google::testing::FakeServer::Handler> >' through '...'
```

Reply:
Even more sadly, clang on MacOS happily compiles this, but crashes at runtime with an invalid instruction... So much for that idea.